repose-aggregator/commons/utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletRequestWrapper.scala

```diff
@@ -0,0 +1,150 @@
+
+  val caseInsensitiveOrdering = Ordering.by[String, String](_.toLowerCase)
+
+  private var headerMap: Map[String, List[String]] = new TreeMap[String, List[String]]()(caseInsensitiveOrdering)
+  private var removedHeaders: Set[String] = new TreeSet[String]()(caseInsensitiveOrdering)
+
+  override def getInputStream: ServletInputStream = {
+    if (status == RequestBodyStatus.Reader) throw new IllegalStateException else status = RequestBodyStatus.InputStream
+    inputStream
+  }
+
+  override def getReader: BufferedReader = {
+    if (status == RequestBodyStatus.InputStream) throw new IllegalStateException else status = RequestBodyStatus.Reader
+    new BufferedReader(new InputStreamReader(inputStream))
+  }
+
+  def getHeaderNamesScala: Set[String] = headerMap.keySet ++ super.getHeaderNames.asScala.toSet.filterNot(removedHeaders.contains)
+
+  override def getHeaderNames: util.Enumeration[String] = getHeaderNamesScala.toIterator.asJavaEnumeration
+
+  override def getHeaderNamesList: util.List[String] = getHeaderNamesScala.toList.asJava
+
+  override def getIntHeader(headerName: String): Int = Option(getHeader(headerName)).getOrElse("-1").toInt
+
+  def getHeadersScala(headerName: String): List[String] = {
+    if (removedHeaders.contains(headerName)) {
+      List[String]()
+    }
+    else {
+      headerMap.getOrElse(headerName, super.getHeaders(headerName).asScala.toList)
+    }
+  }
+
+  override def getHeaders(headerName: String): util.Enumeration[String] = getHeadersScala(headerName).toIterator.asJavaEnumeration
+
+  override def getDateHeader(headerName: String): Long = Option(getHeader(headerName)).map(DateUtils.parseDate(_).getTime).getOrElse(-1)
+
+  override def getHeader(headerName: String): String = getHeadersScala(headerName).headOption.orNull
+
+  override def getHeadersList(headerName: String): util.List[String] = getHeadersScala(headerName).asJava
+
+  override def addHeader(headerName: String, headerValue: String): Unit = {
+    val existingHeaders: List[String] = getHeadersScala(headerName) //this has to be done before we remove from the list,
+                                                                    // because getting this list is partially based on the contents of the removed list
+    if (removedHeaders.contains(headerName)) {
+      removedHeaders = removedHeaders - headerName
+    }
+    headerMap = headerMap + (headerName -> (existingHeaders :+ headerValue))
+  }
+
+  override def addHeader(headerName: String, headerValue: String, quality: Double): Unit = addHeader(headerName, headerValue + ";q=" + quality)
+
+  override def appendHeader(headerName: String, headerValue: String): Unit = {
+    val existingHeaders: List[String] = getHeadersScala(headerName)
+    existingHeaders.headOption match {
+      case Some(value) =>
+        val newHeadValue: String = value + "," + headerValue
+        headerMap = headerMap + (headerName -> (newHeadValue +: existingHeaders.tail))
```

Comment:
Appending the new header value to the end of the first header line messes with the ordering of headers. Would it be better to append the new value to the end of the last header line? Note that ordering DOES matter with headers.